codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Encoder.java

```diff
@@ -190,23 +194,30 @@ private static void encodeInteger(OutputStream out, int mask, int n, int i) thro
     /**
      * Encode string literal according to Section 5.2.
      */
-    private void encodeStringLiteral(OutputStream out, byte[] string) throws IOException {
-        int huffmanLength = Huffman.ENCODER.getEncodedLength(string);
-        if ((huffmanLength < string.length && !forceHuffmanOff) || forceHuffmanOn) {
+    private void encodeStringLiteral(ByteBuf out, CharSequence string) {
+        int huffmanLength = huffmanEncoder.getEncodedLength(string);
+        if ((huffmanLength < string.length() && !forceHuffmanOff) || forceHuffmanOn) {
             encodeInteger(out, 0x80, 7, huffmanLength);
-            Huffman.ENCODER.encode(out, string);
+            huffmanEncoder.encode(out, string);
         } else {
-            encodeInteger(out, 0x00, 7, string.length);
-            out.write(string, 0, string.length);
+            encodeInteger(out, 0x00, 7, string.length());
+            if (string instanceof AsciiString) {
+                // Fast-path
+                AsciiString asciiString = (AsciiString) string;
+                out.writeBytes(asciiString.array(), asciiString.arrayOffset(), asciiString.length());
+            } else {
+                // Only ASCII is allowed in http2 headers, so its fine to use this.
+                // https://tools.ietf.org/html/rfc7540#section-8.1.2
+                out.writeCharSequence(string, CharsetUtil.ISO_8859_1);
```

Comment:
I seem to recall that conversion between charsets in java results in a lot of replacement character (either \uFFFD or '?').  Is that the behavior you want, or did you want an exception thrown?